### ISSUE TITLE
kafka-sink: duplicate project station code with different projectid cause error

### ISSUE THREAD

**Issue by haakondr** [NIVANorge/odm2-postgres-api#196]
We got the following two kafka records:

```
select payload from odm2.kafka_records where payload -> 'data' ->> 'stationCode' = 'E-KOL';
```

```
{
  "time": "2021-11-14T11:40:55.579418",
  "triggeredBy": "niva-data-producer",
  "action": "CREATE",
  "data": {
    "id": 56979,
    "projectId": 8967,
    "stationId": 62868,
    "stationCode": "E-KOL",
    "stationName": "Koller\u00f8dbekken 2",
    "enteredDate": "2020-02-13",
    "auditId": null
  }
}
```

```
{
  "time": "2021-11-14T11:38:08.248892",
  "triggeredBy": "niva-data-producer",
  "action": "CREATE",
  "data": {
    "id": 27637,
    "projectId": 6022,
    "stationId": 62868,
    "stationCode": "E-KOL",
    "stationName": "Koller\u00f8dbekken 2",
    "enteredDate": "2013-12-19",
    "auditId": null
  }
}
```

These two stations have the same station code but are linked to two different project id's. These should be possible to ingest, but we get the following error:

```
Traceback (most recent call last):
  File "/app/kafka_sink/message_handlers.py", line 124, in batch_insert_with_fallback
    await handler(conn, [row])
  File "/app/kafka_sink/message_handlers.py", line 341, in save_project_station_list
    await save_project_station(connection, msg)
  File "/app/kafka_sink/message_handlers.py", line 417, in save_project_station
    await connection.fetchrow(
  File "/usr/local/lib/python3.9/site-packages/asyncpg/connection.py", line 641, in fetchrow
    data = await self._execute(
  File "/usr/local/lib/python3.9/site-packages/asyncpg/connection.py", line 1625, in _execute
    result, _ = await self.__execute(
  File "/usr/local/lib/python3.9/site-packages/asyncpg/connection.py", line 1650, in __execute
    return await self._do_execute(
  File "/usr/local/lib/python3.9/site-packages/asyncpg/connection.py", line 1697, in _do_execute
    result = await executor(stmt, None)
  File "asyncpg/protocol/protocol.pyx", line 201, in bind_execute
asyncpg.exceptions.UniqueViolationError: duplicate key value violates unique constraint "projectstations_projectstationcode_projectid_key"
DETAIL:  Key (projectstationcode, projectid)=(E-KOL, 3320) already exists.
```

Seems like we are attempting to insert to the wrong projectid here? not sure where 3320 is coming from? Are we using a wrong identifier perhaps

**Comment by haakondr** [comment]
update:

there is a third record that comes after which updates the stationcode to `CGL KOL2`:


```
{
  "time": "2022-01-06T09:37:42.187803",
  "triggeredBy": "niva-data-producer",
  "action": "UPDATE",
  "data": {
    "id": 58875,
    "projectId": 12501,
    "stationId": 62868,
    "stationCode": "CGL KOL2",
    "stationName": "Koller\u00f8dbekken 2",
    "enteredDate": "2022-01-06",
    "auditId": null
  }
}
```

Looking up the projectstation, it seems like it was updated correctly:

```
select ps.*, e.projectstationexternalidentifier
from odm2.projectstationexternalidentifiers e
         inner join odm2.projectstations ps on ps.projectstationid = e.projectstationid
where e.projectstationexternalidentifier = '58875';
```

| projectstationid | projectstationcode | projectstationname | projectid | samplingfeatureid | projectstationexternalidentifier |
| :--- | :--- | :--- | :--- | :--- | :--- |
| 123084 | CGL KOL2 | Kollerødbekken 2 | 3320 | 1 | 58875 |


Summarized, it seems like it is working, but the sequence of events caused a duplicate code error before everything was ingested. With the following order:


| type | data |
| :--- | :--- |
| "CREATE" | {"id": 56952, "projectId": 12500, "stationId": 62868, "stationCode": "CGL KOL2", "stationName": "Koller\\u00f8dbekken 2, Dalebekken", "enteredDate": "2020-02-05", "auditId": null} |
| "CREATE" | {"id": 56979, "projectId": 8967, "stationId": 62868, "stationCode": "E-KOL", "stationName": "Koller\\u00f8dbekken 2", "enteredDate": "2020-02-13", "auditId": null} |
| "INSERT" | this is not present in kafka_records table as we update it, but we must have had an insert here with stationcode = 'E-KOL'. and projectstation id = 58875. This is the one that failed | 
| "UPDATE" | {"id": 58875, "projectId": 12501, "stationId": 62868, "stationCode": "CGL KOL2", "stationName": "Koller\\u00f8dbekken 2", "enteredDate": "2022-01-06", "auditId": null} |
| "CREATE" | {"id": 27637, "projectId": 6022, "stationId": 62868, "stationCode": "E-KOL", "stationName": "Koller\\u00f8dbekken 2", "enteredDate": "2013-12-19", "auditId": null} |

**Comment by haakondr** [comment]
it seems like our topics are not compacted the way I thought they would be. I think we can tune this by configuring `segment.bytes`. The caveat here is that a topic is not compacted for an active "segment". We could decrease our segment size, which would to some degree fix this, but we could always get into a situation where a row is updated right after insert and possibly not trigger compaction before consumption.

In other words, we should probably deal with this consumer side anyway.

Suggested solution:

- we filter our list of messages on poll and only proceed with the most recent of each key
